### DISCUSSION TITLE
When auth with google I got error - missing sub claim - status code 401

### DISCUSSION THREAD

**Discussion by nicitaacom** [supabase/supabase#16658] [Questions] [answered]
I want
```tsx
 await supabase.auth.signInWithOAuth({ provider: "google" })
    const response = await supabase.auth.getSession()
```

And than set response data to my store (to save it in localstorage to don't do every time request to backend)
 
I use React + Vite + TypeScript + tailwind + zustand + supabase
 
I tried do it driectly in store but got the same error
 
 ```ts
 
export const authUserWithProvider = async () => {
  await supabase.auth.signInWithOAuth({ provider: "google" })
  const response = await supabase.auth.getUser()
  return response && response
}

```

I called this function like this
```tsx
              <Button variant="continue-with" onClick={userStore.authUserWithProvider}>

```

Is it possible in loginWithGoogle() function after login with google do something with getUser() data?
Now I even don't see any console.log - because page refreshes
```ts
  async function loginWithGoogle() {
    await supabase.auth.signInWithOAuth({ provider: "google" })
    const response = await supabase.auth.getSession()
    setTimeout(() => {
      console.log("response - ", response)
    }, 1000)
    console.log(105, "response - ", response)
    console.log(106, "response.data - ", response.data)
  }
  ```

**Comment by GaryAustin1** [comment]
signInWithOAuth() does not return or set a session after the call.  It will redirect back to your app either from siteURL or any redirectTo url you have in the call options.   So after you signin with the provider your app should restart again based on that URL.  The return from signInWithOAuth is just an error or the URL that was used to get to the provider.

**Reply by nicitaacom** [reply]
Tell me please more about parameters
Where I should define it in my onClick event or here on site?

![image](https://github.com/supabase/supabase/assets/39565703/23087e93-d153-472c-a2db-3309553e8607)


I have good idea - redirect user to route /oauth with useEffect that set data from getUser() to local storage and when it's done redirect user back to /
It fix 2 issues - 1 about /# in the url after login with OAuth and 2 - about saving data to localstorage (to don't do every time request to backend)

How do I redirect user to /oauth route after login with google?

**Comment by nicitaacom** [comment, marked as answer]
Auth flow with providers works differently - you may achieve this effect like below (to claim user data for future use)
After OAuth I should redirect user to protected route like auth/callback/client like so
```ts
 const { error } = await supabaseClient.auth.signInWithOAuth({
        provider: "google",
        options: { redirectTo: `${location.origin}/auth/callback/` },
      })
      if (error) throw error
  ```
  
  On this route insert rows in tabe users (because we don't know users login in first time or this row in table users already exists)
  
So when I redirect user to auth/callback in route.ts I do this

**auth/callback/route.ts**
```ts
import { createRouteHandlerClient } from "@supabase/auth-helpers-nextjs"
import { cookies } from "next/headers"
import { NextResponse } from "next/server"

export async function GET(request: Request) {
  const requestUrl = new URL(request.url)
  const code = requestUrl.searchParams.get("code")

  const supabase = createRouteHandlerClient({ cookies })
  if (code) {
    await supabase.auth.exchangeCodeForSession(code)
  }
  const { data: user } = await supabase.auth.getUser()
  if (user && user.user) {
    //Insert row in users table
    const { error: error_insert_users } = await supabase.from("users").insert({
      id: user.user.id,
      username: user.user.user_metadata.name,
      email: user.user.user_metadata.email,
      profile_picture_url: user.user.user_metadata.picture,
    })
    //Insert row in users_cart table
    const { error: error_users_cart_insert } = await supabase
      .from("users_cart")
      .insert({ owner_username: user.user.user_metadata.name, id: user.user.id })
  }
  return NextResponse.redirect(
    `${requestUrl.origin}/auth/client?userId=${user?.user?.id}&username=${user.user?.user_metadata.name}
    &email=${user.user?.user_metadata.email}&profile_picture_url=${user.user?.user_metadata.picture}`,
  )
}

```

Then I insert rows in table users with user info

  **auth/callback/client/page.tsx**
  ```tsx
  "use client"
import useUserStore from "@/store/user/userStore"
import { useRouter } from "next/navigation"
import { useSearchParams } from "next/navigation"

export default function Page() {
  const userStore = useUserStore()
  const router = useRouter()

  const userId = useSearchParams().get("userId")
  const username = useSearchParams().get("username")
  const email = useSearchParams().get("email")
  const profile_picture_url = useSearchParams().get("profile_picture_url")
  userStore.setUser(userId ?? "", username ?? "", email ?? "", profile_picture_url ?? "")
  router.push("/")
}

```
